Implement ExtEvLoop (PECL ext-ev)

kaduev13 opened the pull request:
@clue, @WyriHaximus, @jsor hello!

Finally I fixed my old [pull request](https://github.com/reactphp/event-loop/pull/97). I'm really sorry that it took so long for me.

I hope, that this feature still wanted and needed :)

WyriHaximus commented:
Hey @kaduev13 yes your PR is still wanted, in fact the plan was to take over and finish your PR once `v0.5.0` gone out. Will look into this one later today or tomorrow :+1: .

clue commented on src/ExtEvLoop.php:
```diff
@@ -0,0 +1,221 @@
+            },
+            function ($signal) {
+                if ($this->signals->count($signal) === 0) {
+                    unset($this->signalEvents[$signal]);
```
This should probably `stop()` the watcher before clearing this reference, no?

kaduev13 replied:
Yes, it's very important to stop watcher before clearing the reference. Fixed.

clue commented on src/ExtEvLoop.php:
```diff
@@ -0,0 +1,221 @@
+    private function getStreamListenerClosure($stream, callable $listener)
+    {
+        return function () use ($stream, $listener) {
+            call_user_func($listener, $stream, $this);
```
The additional `$loop` argument has been removed via #110, so the last parameter should be omitted here.

kaduev13 replied:
Thank you for pointing it! I removed this argument.

clue commented on src/Factory.php:
```diff
@@ -26,7 +26,10 @@ public static function create()
+        } elseif (class_exists('EvLoop', false)) {
+            return new ExtEvLoop();
+        }
+        elseif (class_exists('EventBase', false)) {
```
Minor CS issue.

kaduev13 replied:
Fixed, thank you :)

clue commented on tests/AbstractLoopTest.php:
```diff
@@ -534,12 +534,12 @@ public function testSignalsKeepTheLoopRunning()
         });
 
-        $this->assertRunSlowerThan(1.5);
+        $this->assertRunSlowerThan(2);
```
What is the motivation for this change?

kaduev13 replied:
I thought that there was a problem with floating point accuracy, but it was not. It should be changed to the previous `1.5` value. Fixed.

kaduev13 commented:
Also, I noticed that other loops implementations contains such code for signals callbacks:
```php
function ($signal) {
    $this->signalEvents[$signal] = new SignalEvent($f = function () use ($signal, &$f) {
        $this->signals->call($signal);
        // Ensure there are two copies of the callable around until it has been executed.
        // For more information see: https://bugs.php.net/bug.php?id=62452
        // Only an issue for PHP 5, this hack can be removed once PHP 5 support has been dropped.
        $g = $f;
        $f = $g;
    }, $signal);
    $this->loop->add($this->signalEvents[$signal]);
},
```

My question is why using the `$f` identifier? Why not just
```php
function ($signal) {
    $this->signalEvents[$signal] = new SignalEvent(function () use ($signal) {
        $this->signals->call($signal);
    }, $signal);
    $this->loop->add($this->signalEvents[$signal]);
}
```

kaduev13 commented:
@WyriHaximus, I checked that issue, but, as I understood, this bug is about aliasing. I mean, that this code will randomly fail:
```php
function ($signal) {
    $this->signalEvents[$signal] = new SignalEvent($f = function () use ($signal, &$f) {
        $this->signals->call($signal);
    }, $signal);
    $this->loop->add($this->signalEvents[$signal]);
}
```
but this one should not (no alias == no problems?):
```php
function ($signal) {
    $this->signalEvents[$signal] = new SignalEvent(function () use ($signal) {
        $this->signals->call($signal);
    }, $signal);
    $this->loop->add($this->signalEvents[$signal]);
}
```

WyriHaximus commented:
@kaduev13 feel free to prove me wrong, I don't like the solution for that bug so if we can get rid of it that would be great :+1: . Also be sure to check signal example to be sure. IIRC this is one of the failing builds for that bug: https://travis-ci.org/reactphp/event-loop/jobs/258431073#L1054

kaduev13 commented:
@WyriHaximus, thank you! Unfortunately, the build url you provided is for commit, that does not [exist](https://github.com/reactphp/event-loop/commit/f5e6f174871102794934df8b65b3c4dff36c9f0a). I'll try to reproduce this behavior.

WyriHaximus commented:
@kaduev13 yes correct/sorry about that the commits where squashed just before merging to keep a clean commit log.

kaduev13 commented:
@WyriHaximus sorry, I didn't have enough time to completely test the issue with lambdas. I'll continue my investigation later. I added similar workaround to `ExtEvLoop` to ensure that everything will work well.
@clue, I fixed issues you have pointed at, review this changes again, please :)

clue commented:
@kaduev13 Very good point about signal handling callbacks! I've just filed #150, once this is in, this should be simplified significantly (this will likely cause a merge conflict here though).

kaduev13 commented:
@clue nice work with signals, thank you! I rebased this branch onto master, review pls =)

clue commented on src/ExtEvLoop.php:
```diff
@@ -0,0 +1,255 @@
+    public function isTimerActive(TimerInterface $timer)
+    {
+        return $this->timers->contains($timer);
+    }
```
This method has been removed from the public interface with #133, so this should probably be removed here.

clue commented on README.md:
```diff
@@ -201,6 +201,16 @@ It supports the same backends as libevent.
 
 This loop is known to work with PHP 5.4 through PHP 7+.
 
+#### ExtEvLoop
```
Should also be added to the TOC above :+1:

kaduev13 commented:
@clue thank you for the review :) I fixed these issues.

WyriHaximus commented:
Awesome, thank you 👍 !

kaduev13 commented:
Great to know that it's merged! Thank you!

WyriHaximus commented:
@kaduev13 [`v0.5.1`](https://github.com/reactphp/event-loop/releases/tag/v0.5.1)

WyriHaximus commented:
@kaduev13 FYI I've been running it in production for a week now, as expected no issue what so ever. Well done 👍
